File: LeetCode/python/SumRootToLeafNumbers.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sumLst(self,lst):
        lstSize = len(lst)
        sum = 0
        for i in range(0,lstSize):
            tmp = lst[i]
            while(i<lstSize-1):
                tmp*=10
                i+=1
                
            sum += tmp
            
        return sum
    
    def sumHelper(self,root,lst,sum):
        if(root != None):
            if(root.left == None and root.right == None):
                tmplst = []
                for i in lst:
                    tmplst.append(i)
                
                tmplst.append(root.val)
                sumNum = self.sumLst(tmplst)
                print(sumNum)
                sum[0] += sumNum
            else: 
                tmpLst = []
                for i in lst:
                    tmpLst.append(i)
                    
                tmpLst.append(root.val)
                      
                if(root.left != None):
                    self.sumHelper(root.left,tmpLst,sum)
                
                if(root.right != None):
                    self.sumHelper(root.right,tmpLst,sum)

    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        sumNum = [0]
        self.sumHelper(root,[],sumNum)
        return sumNum[0]

        
#测试程序
root = TreeNode(1)
node1 = TreeNode(2)
node2 = TreeNode(3)

node3 = TreeNode(4)
node4 = TreeNode(5)
node5 = TreeNode(6)

root.left = node1
root.right = node2

node1.left = node3
node1.right = node4

node2.left = node5

s = Solution()
print(s.sumNumbers(root))

File: LeetCode/iter2/C++/SumRootToLeafNumbers.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:

	int sumNumbers(TreeNode* root) {
		sum = 0;

		if (root != NULL)
		{
			string s = "";
			s += char(root->val + '0');

			if(root->left != NULL)
				helper(root->left,s);

			if(root->right != NULL)
				helper(root->right,s);

			if (root->left == NULL && root->right == NULL)
				sum += str2int(s);
		}

		return sum;
	}

private:
	int sum;

	void helper(TreeNode* root, string s)
	{
		if (root->left == NULL  && root->right == NULL)
		{
			string tmpS = s + char(root->val + '0');
			sum += str2int(tmpS);
		}
		else
		{
			string tmpS = s + char(root->val + '0');

			if (root->left != NULL)
				helper(root->left, tmpS);

			if (root->right != NULL)
				helper(root->right, tmpS);
		}
	}

	int str2int(string s)
	{
		stringstream ss;
		ss << s;
		int num;
		ss >> num;

		return num;
	}
};

int main()
{
	TreeNode *root = new TreeNode(1);
	TreeNode *node1 = new TreeNode(2);
	TreeNode *node2 = new TreeNode(3);

	root->left = node1;
	node1->right = node2;

	Solution s;

	cout << s.sumNumbers(root);


	return 0;
}